app/Singletons/Mods.php

```diff
@@ -135,19 +135,31 @@ public function idsToBitset($ids): int
         );
     }
 
-    public function assertValidExclusivity(int $rulesetId, array $requiredIds, array $allowedIds): bool
+    public function assertValidExclusivity(int $rulesetId, array $modAcronyms): bool
     {
+        while (($modAcronym = array_pop($modAcronyms)) !== null) {
+            $mod = $this->mods[$rulesetId][$modAcronym];
+            $incompatibleIds = $mod['IncompatibleMods'];
+
+            $invalidIds = $incompatibleIds->intersect(new Set($modAcronyms));
+            if ($invalidIds->count() > 0) {
+                throw new InvariantException("incompatible mods: {$modAcronym}, {$invalidIds->join(', ')}");
+            }
+        }
+
+        return true;
```

Comment:
this and the other function probably shouldn't return anything. They're assertions.

Reply:
Applied in https://github.com/ppy/osu-web/pull/11875/commits/47289670275a8639a48b4c3f4d41ade428e82a65; this affected tests too because they were checking the retval.